# OOP- Object Oriented Programming

In [1]:
class NameOfClass():
    def _init_(self,param1,param2):
        self.param1=param1
        self.param2=param2
        
    def some_method(self):
        #perform some action
        print (self.param1)

#Started

In [19]:
mylist = [1,2,3]
myset = set(mylist)
mytuple = tuple(mylist)

In [11]:
type(mylist)

list

In [12]:
type(myset)

set

In [13]:
myset

{1, 2, 3}

In [14]:
mytuple

(1, 2, 3)

In [15]:
mylist

[1, 2, 3]

In [160]:
class Dog():
    
    def __init__(self,breed): ##has to be 2 underscores
        
        self.breed = breed

In [161]:
my_dog = Dog()

TypeError: Dog.__init__() missing 1 required positional argument: 'breed'

In [54]:
my_dog = Dog(breed="Lab")

In [55]:
type(my_dog)

__main__.Dog

In [56]:
my_dog.breed

'Lab'

In [156]:
class Dog():
    
    def __init__(self,mybreed): ##has to be 2 underscores # this is now a "method" instead of a "function"
        
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = mybreed

In [157]:
my_dog = Dog( "Huskie")

In [158]:
type(my_dog)

__main__.Dog

In [159]:
my_dog.breed

'Huskie'

In [9]:
class Dog():
    
    def __init__(self,mybreed): ##has to be 2 underscores # this is now a "method" instead of a "function"
        
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.my_attribute = mybreed

In [10]:
my_dog = Dog(mybreed = "Huskie")

In [11]:
type(my_dog)

__main__.Dog

In [12]:
my_dog.my_attribute

'Huskie'

In [186]:
class Dog():
    
    def __init__(self,breed,name,spots): ##has to be 2 underscores # this is now a "method" instead of a "function"
                    #right now breed = string, name = string, spots = string but in this we are taking as either True or False
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = breed
        self.name = name
        
        #expect the boolean True or False
        self.spots = spots

In [187]:
my_dog = Dog(breed="lab",name="Sammy",spots= "No spots")

In [188]:
type(my_dog)

__main__.Dog

In [189]:
my_dog.breed

'lab'

In [190]:
my_dog.name

'Sammy'

In [191]:
print(my_dog.name)

Sammy


In [25]:
my_dog.spots

'No spots'

In [13]:
class Dog():
    
    #CLASS OBJECT ATTRIBUTE
    #SAME FOR ANY INSTANCE OF A CLASS
    species = "mammal" #this attribute has to be above init 
    
    def __init__(self,breed,name,spots): #has to be 2 underscores # this is now a "method" instead of a "function" #self is a keyword
                    #right now breed = string, name = string, spots = string but in this we are taking as either True or False
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = breed
        self.name = name
        
        #expect the boolean True or False
        self.spots = spots

In [14]:
my_dog = Dog("Lab","Sam","False")

In [15]:
type(my_dog)

__main__.Dog

In [16]:
my_dog.species

'mammal'

In [17]:
my_dog.name

'Sam'

In [18]:
class Dog():
    
    #CLASS OBJECT ATTRIBUTE
    #SAME FOR ANY INSTANCE OF A CLASS
    species = "mammal" #this attribute has to be above init 
    
    def __init__(self,breed,name): #has to be 2 underscores # this is now a "method" instead of a "function" #self is a keyword
                    #right now breed = string, name = string, spots = string but in this we are taking as either True or False
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = breed
        self.name = name
        
        #expect the boolean True or False
        #self.spots = spots
        
    #OPERATIONS/ACTIONS ----> Methods
    def bark(self):
        print("Woof!!!")

In [19]:
my_dog = Dog ("Lab","Frankie")

In [20]:
type(my_dog)

__main__.Dog

In [21]:
my_dog.species

'mammal'

In [22]:
my_dog.name

'Frankie'

In [24]:
my_dog.bark()

Woof!!!


In [42]:
class Dog():
    
    #CLASS OBJECT ATTRIBUTE
    #SAME FOR ANY INSTANCE OF A CLASS
    species = "mammal" #this attribute has to be above init 
    
    def __init__(self,breed,name): #has to be 2 underscores # this is now a "method" instead of a "function" #self is a keyword
                    #right now breed = string, name = string, spots = string but in this we are taking as either True or False
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = breed
        self.name = name
        
        #expect the boolean True or False
        #self.spots = spots
        
    #OPERATIONS/ACTIONS ----> Methods
    def bark(self):
        print("Woof!!! My name is {}".format(self.name))

In [27]:
my_dog = Dog ("Lab","Frankie")

In [28]:
type(my_dog)

__main__.Dog

In [29]:
my_dog.species

'mammal'

In [30]:
my_dog.name

'Frankie'

In [31]:
my_dog.bark()

Woof!!! My name is Frankie


In [250]:
class Dog():
    
    #CLASS OBJECT ATTRIBUTE
    #SAME FOR ANY INSTANCE OF A CLASS
    species = "mammal" #this attribute has to be above init 
    
    def __init__(self,breed,name): #has to be 2 underscores # this is now a "method" instead of a "function" #self is a keyword
                    #right now breed = string, name = string, spots = string but in this we are taking as either True or False
        #Attributes
        #We take in the argument
        #Assign it using self.attribute_name
        
        self.breed = breed
        self.name = name
        
        #expect the boolean True or False
        #self.spots = spots
        
    #OPERATIONS/ACTIONS ----> Methods
    def bark(self,number):
        print("Woof!!! My name is {} and the number is {}".format(self.name, number))# here we are not using self.number because
                                                       #number us already been provided when we call bark()

In [251]:

my_dog = Dog ("Lab","Frankie")

In [252]:
my_dog.bark(10)

Woof!!! My name is Frankie and the number is 10


In [312]:
class Circle():
    
    #circle attribute
    pi = 3.14
    
    def __init__(self,radius=1):
        
        self.radius = radius
        self.area = radius**2*self.pi
        
    #method
    def get_circum(self):
        return 2*self.pi*self.radius

In [313]:
a_circle = Circle(30)

In [314]:
a_circle.pi

3.14

In [315]:
a_circle.radius 

30

In [316]:
a_circle.get_circum()

188.4

In [97]:
a_circle.area

2826.0

In [106]:
class Animal():
    
    
    def __init__(self):
        print ("ANIMAL CREATED")
        
    def who_am_i(self):
        print ("I am an animal")
        
    def eat(self):
        print("I am eating")

In [111]:
myanimal = Animal()

ANIMAL CREATED


In [108]:
myanimal.who_am_i()

I am an animal


In [109]:
myanimal.eat()

I am eating


In [130]:
class Dog(Animal): #Animal class will serve as a base class to the New created Dog class and dog class inhertited or derive 
                   #the Animal class for some methods
    
    def __init__(self):
        Animal.__init__(self) #inherting the animal class. This will help to take the values from Animal method
        print ("Dog Created")
        
    def who_am_i(self): #when you want to overwrite some method then you have to just re-write the method again #thats-it
        print ("I am a dog!!!")

In [127]:
mydog = Dog()

ANIMAL CREATED
Dog Created


In [128]:
mydog.eat()

I am eating


In [129]:
mydog.who_am_i()

I am a dog!!!


In [132]:
class Dog(Animal): #Animal class will serve as a base class to the New created Dog class and dog class inhertited or derive 
                   #the Animal class for some methods
    
    def __init__(self):
        Animal.__init__(self) #inherting the animal class. This will help to take the values from Animal method
        print ("Dog Created")
        
    #def who_am_i(self): #when you want to overwrite some method then you have to just re-write the method again #thats-it
     #   print ("I am a dog!!!")
    
    def eat(self):
        print("I am a dog and eating")
        
    def bark(self):
        print("Woof!!!")

In [135]:
adog = Dog()

ANIMAL CREATED
Dog Created


In [137]:
adog.eat()

I am a dog and eating


In [138]:
adog.who_am_i()

I am an animal


In [140]:
adog.bark()

Woof!!!


# Polymorphism

In [174]:
class Dog():
    
    def __init__(self,name):
        self.name = name
        
    def speak(self):
        return self.name + " says woof!!!"

In [175]:
class Cat():
    
    def __init__(self,name):
        self.name = name
        
    def speak(self):
        return self.name + " says meow!!!"

In [182]:
niko = Dog("niko")
felix = Cat("felix")

In [183]:
print (niko.speak())
print (felix.speak())

niko says woof!!!
felix says meow!!!


In [204]:
niko.name

'niko'

In [205]:
type(niko)

__main__.Dog

In [195]:
niko.name()

TypeError: 'str' object is not callable

In [202]:
print(niko.name)

niko


In [212]:
class Dog():
    
    def __init__(self,name):
        self.name = name
        
    def speak(self):
        return self.name + " says woof!!!"

In [213]:
class Cat():
    
    def __init__(self,name):
        self.name = name
        
    def speak(self):
        return self.name + " says meow!!!"

In [ ]:
for pet in [niko,felix]:
    print(type(pet))
    print(pet.speak())

In [214]:
def pet_speak(pet):
    print (pet.speak())

In [215]:
pet_speak(niko)

niko says woof!!!


In [216]:
pet_speak(felix)

felix says meow!!!


In [224]:
class Animal():
    
    def __init__(self,name):
        self.name = name
        
    def speak(self):
        raise NotImplementedError("Subclass must implement this abstract method")

In [225]:
myanimal = Animal("fred")

In [226]:
myanimal.speak()

NotImplementedError: Subclass must implement this abstract method

class Dog(Animal): #Animal class will serve as a base class to the New created Dog class and dog class inhertited or derive 
                   #the Animal class for some methods
    
    def __init__(self):
        Animal.__init__(self) #inherting the animal class. This will help to take the values from Animal method
        print ("Dog Created")

In [261]:
class Dog(Animals):
    
    def speak(self):
        return self.name+" says woof!!!"

In [262]:
class Cat(Animals):
    
    def speak(self):
        return self.name+" says meow!!!"

In [263]:
fido = Dog("fido")
isis = Cat("isis")

In [253]:
def calc_of_area(shape):
    return shape.area()
    
class Square():
        
    def __init__(self,side):
        self.side = side
        
    def area(self):
        return self.side**2
            
class Cirlce():
            
    def __init__(self,radius):
        self.radius = radius
                
    def area(self):
        pi = 3.14
        return self.pi*(self.radius**2)
        

In [254]:
square = Square(2)
circle = Circle(3)

In [255]:
print (calc_of_area(square))
print (calc_of_area(circle))

4
28.259999999999998


def calculate_area(shape):
    return shape.area()  # Relies on the object having an "area()" method

class Square:
    def __init__(self, side):
        self.side = side

    def area(self):
        return self.side * self.side

class Circle:
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3.14 * self.radius * self.radius

square = Square(5)
circle = Circle(3)

print(calculate_area(square))  # Outputs 25
print(calculate_area(circle))  # Outputs 28.26


#Special Method / Magic Method

In [264]:
mylist = [1,2,3]

In [265]:
len(mylist)

3

In [266]:
class Sample():
    pass

In [267]:
mysample = Sample()

In [268]:
print (mysample)

In [270]:
print (mylist)

[1, 2, 3]


In [271]:
class Book():
    
    def __init__ (self, title,author,pages):
        self.title = title
        self.author = author
        self.pages = pages
        


In [272]:
b = Book("Python rocks","Jose", 200)
print(b) #this happens because this b asks for the string representation

In [273]:
str(b) #change to string

'<__main__.Book object at 0x0000022BECD46490>'

In [277]:
class Book():
    
    def __init__ (self, title,author,pages):
        self.title = title
        self.author = author
        self.pages = pages
        
    def __str__(self):
        return f"{self.title} by {self.author}"

In [278]:
b = Book("Python rocks","Jose", 200)
print(b) 

Python rocks by Jose


In [279]:
str(b)

'Python rocks by Jose'

In [280]:
len(b)

TypeError: object of type 'Book' has no len()

In [288]:
class Book():
    
    def __init__ (self, title,author,pages):
        self.title = title
        self.author = author
        self.pages = pages
        
    def __str__(self):
        return f"{self.title} by {self.author}"
    
    def __len__(self):
        return self.pages
    

In [296]:
b = Book("Python rocks","Jose",20.10)
print(b) 

Python rocks by Jose


In [297]:
len(b) #won't take any float value

TypeError: 'float' object cannot be interpreted as an integer

In [303]:
b = Book("Python rocks","Jose",200)
print(b) 

Python rocks by Jose


In [304]:
len(b)

200

In [305]:
del b #with the help of the "del" keyword you can delete variable from system's memory

In [306]:
b

NameError: name 'b' is not defined

In [368]:
class Book():
    
    def __init__ (self, title,author,pages):
        self.title = title
        self.author = author
        self.pages = pages
        
    def __str__(self):
        return f"{self.title} by {self.author}"
    
    def __len__(self):
        return self.title
    def __del__(self):
        pass #print("A book object has been deleted")

In [369]:
b = Book("Python rocks","Jose",20.10)
print(b) 

A book object has been deleted
Python rocks by Jose


In [370]:
del b

In [371]:
b

NameError: name 'b' is not defined

In [365]:
len(b)

TypeError: 'str' object cannot be interpreted as an integer

.

In [351]:
class Simple():
    
    def __init__ (self,value):
        self.value = value
        
    def add_to_value(self,amount):
        self.value = self.value + amount
        
    def __str__(Simple):
        return f"{self.value} \n{self.value}"

In [352]:
simple = Simple(3000)

In [353]:
simple.value

3000

In [354]:
simple.add_to_value(500)

In [355]:
simple.value

3500

In [356]:
print(simple)

NameError: name 'self' is not defined